### ISSUE TITLE
Compilation fails with GEGL + introspection + as-needed + GNU gold linker

### ISSUE THREAD

**Issue by hartwork** [mypaint/libmypaint#58]
Hi there!

I would like to forward-report a [libmypaint compilation/linking error](https://bugs.gentoo.org/show_bug.cgi?id=589266), that I have not been able to fix myself.
To reproduce it, these things are needed together:
- Enabling support for both GEGL and introspection
- `LDFLAGS="-Wl,--as-needed"`
- Linkling with `ld.gold` rather than `ld.bfd`

Below I am doing just that on a shell.
The error is `libmypaint-gegl.so.0: undefined symbol: mypaint_surface_unref`, a few lines up the end.

I'm aware that combination may not be what everyone is doing but I believe it makes more sense to report it to you rather than just working around symptoms on our end until someone runs into it somewhere else.

Best, Sebastian

```
# sudo binutils-config --linker ld.gold
# ld --version | head -n 1
GNU gold (Gentoo 2.25.1 p1.1 2.25.1) 1.11

# git clone https://github.com/mypaint/libmypaint.git
# cd libmypaint/
# ./autogen.sh
# ./configure LDFLAGS="-Wl,--as-needed" --enable-introspection --enable-gegl
# make
[..]
make[2]: Entering directory '/mydir/libmypaint/gegl'
CPPFLAGS="" CFLAGS="-D_POSIX_C_SOURCE=200809L -g -O2" LDFLAGS="-Wl,--as-needed" CC="gcc -std=gnu99" PKG_CONFIG="/usr/bin/pkg-config" DLLTOOL="false" CFLAGS="-D_POSIX_C_SOURCE=200809L -g -O2" LDFLAGS="-Wl,--as-needed -lm" CPPFLAGS="" CXXFLAGS="" /usr/bin/g-ir-scanner  --warn-all --pkg="gegl-0.3" --pkg="glib-2.0" --namespace="MyPaintGegl" --nsversion="1.3" --identifier-prefix="MyPaintGegl" --symbol-prefix="mypaint_gegl" --add-include-path="./.." --add-include-path="." --add-include-path="./.." --add-include-path="`/usr/bin/pkg-config --define-variable=datadir="/usr/local/share" --variable=girdir gobject-introspection-1.0`" --namespace=MyPaintGegl --nsversion=1.3 --libtool="/bin/sh ../libtool"  --include=GObject-2.0 --include=MyPaint-1.3 --include=Gegl-0.3   --library=libmypaint-gegl.la --library=../libmypaint.la  --cflags-begin  -I.. -I.. -I. -I.. --cflags-end  ../glib/mypaint-gegl-glib.h mypaint-gegl-surface.h ../glib/mypaint-gegl-glib.c mypaint-gegl-surface.c libmypaint-gegl.la Makefile --output MyPaintGegl-1.3.gir
g-ir-scanner: link: /bin/sh ../libtool --mode=link --tag=CC gcc -std=gnu99 -o /mydir/libmypaint/gegl/tmp-introspectjidC3k/MyPaintGegl-1.3 -export-dynamic -D_POSIX_C_SOURCE=200809L -g -O2 -Wl,--as-needed -lm tmp-introspectjidC3k/mydir/libmypaint/gegl/tmp-introspectjidC3k/MyPaintGegl-1.3.o -L. libmypaint-gegl.la ../libmypaint.la -lgegl-0.3 -lgegl-npd-0.3 -lm -Wl,--export-dynamic -lgmodule-2.0 -pthread -ljson-glib-1.0 -lgio-2.0 -lgobject-2.0 -lbabl-0.1 -lglib-2.0
libtool: link: gcc -std=gnu99 -o /mydir/libmypaint/gegl/tmp-introspectjidC3k/.libs/MyPaintGegl-1.3 -D_POSIX_C_SOURCE=200809L -g -O2 -Wl,--as-needed tmp-introspectjidC3k/mydir/libmypaint/gegl/tmp-introspectjidC3k/MyPaintGegl-1.3.o -Wl,--export-dynamic -pthread -Wl,--export-dynamic  -L. ./.libs/libmypaint-gegl.so ../.libs/libmypaint.so -ljson-c -lgegl-0.3 -lgegl-npd-0.3 -lm -lgmodule-2.0 -ljson-glib-1.0 -lgio-2.0 -lgobject-2.0 -lbabl-0.1 -lglib-2.0 -pthread
/mydir/libmypaint/gegl/tmp-introspectjidC3k/.libs/MyPaintGegl-1.3: symbol lookup error: /mydir/libmypaint/gegl/.libs/libmypaint-gegl.so.0: undefined symbol: mypaint_surface_unref
Command '[u'/mydir/libmypaint/gegl/tmp-introspectjidC3k/MyPaintGegl-1.3', u'--introspect-dump=/mydir/libmypaint/gegl/tmp-introspectjidC3k/functions.txt,/mydir/libmypaint/gegl/tmp-introspectjidC3k/dump.xml']' returned non-zero exit status 127
make[2]: *** [/usr/share/gobject-introspection-1.0/Makefile.introspection:156: MyPaintGegl-1.3.gir] Error 1
make[2]: Leaving directory '/mydir/libmypaint/gegl'
make[1]: *** [Makefile:822: all-recursive] Error 1
make[1]: Leaving directory '/mydir/libmypaint'
make: *** [Makefile:579: all] Error 2
```

**Comment by Jehan** [comment]
I can reproduce indeed a link failure during libmypaint-gegl introspection with ld.gold and the link option --as-needed.
By the way `binutils-config` does not exist on my machine. I assume that's a distribution specific tool. It seems that the equivalent standard way to change the linker would be with CFLAGS="-fuse-ld=gold", right?

Anyway I get a similar error except that the undefined symbol is `mypaint_surface_ref` here:

> /home/jehan/dev/build/libmypaint/gegl/tmp-introspect9pXA2W/.libs/MyPaintGegl-1.3: symbol lookup error: /home/jehan/dev/build/libmypaint/gegl/.libs/libmypaint-gegl.so.0: undefined symbol: mypaint_surface_ref

Now I could not understand the source cause. I tried to play with the options of the command that `make` is running but did not find what is wrong. Also I am far from being an introspection expert so I don't get all the details of what does this `g-ir-scanner` command line which breaks (to be run from the `gegl/` subdir):

> CC="gcc" CFLAGS="-D_POSIX_C_SOURCE=200809L -fuse-ld=gold" LDFLAGS="-Wl,--as-needed -lm" CPPFLAGS="" CXXFLAGS="" /usr/bin/g-ir-scanner --warn-all --pkg="gegl-0.3" --pkg="glib-2.0" --namespace="MyPaintGegl" --nsversion="1.3" --identifier-prefix="MyPaintGegl" --symbol-prefix="mypaint_gegl" --add-include-path="/path/to/src/libmypaint/gegl/.." --add-include-path="/path/to/src/libmypaint/gegl" --add-include-path="./.." --add-include-path="`/usr/bin/pkg-config --define-variable=datadir="/usr/local/share" --variable=girdir gobject-introspection-1.0`" --namespace=MyPaintGegl --nsversion=1.3 --libtool="/bin/sh ../libtool"  --include=GObject-2.0 --include=MyPaint-1.3 --include=Gegl-0.3   --library=libmypaint-gegl.la --library=../libmypaint.la --cflags-begin  -I.. -I/path/to/src/libmypaint -I. -I.. -lmypaint --cflags-end  /path/to/src/libmypaint/gegl/../glib/mypaint-gegl-glib.h /path/to/src/libmypaint/gegl/mypaint-gegl-surface.h /path/to/src/libmypaint/gegl/../glib/mypaint-gegl-glib.c /path/to/src/libmypaint/gegl/mypaint-gegl-surface.c libmypaint-gegl.la ../libmypaint.la Makefile --output MyPaintGegl-1.3.gir

As soon as you remove "-fuse-ld=gold" or "Wl,--as-needed", it works. So basically if anyone figures out how to run this command successfully with these 2 options, we should be able to fix the Makefile. I spent 30 min on it and did not find any solution.

Also it may simply be a bug in ld.gold. On the web I found similar issues:
https://bugs.archlinux.org/task/40515

**Comment by hartwork** [comment]
> By the way `binutils-config` does not exist on my machine. I assume that's a distribution specific tool. It seems that the equivalent standard way to change the linker would be with CFLAGS="-fuse-ld=gold", right?

`-fuse-ld=gold` didn't work for me when I tried it first but it seems I can confirm now that
```
# ./configure CFLAGS="-fuse-ld=gold" LDFLAGS="-Wl,--as-needed" --enable-introspection --enable-gegl
# make clean all
```
works to trigger the bug.

> Also it may simply be a bug in ld.gold. On the web I found similar issues:
> https://bugs.archlinux.org/task/40515

Maybe.

What I don't get is why despite line
```
libmypaint_gegl_la_LIBS = $(GEGL_LIBS)
```
or
```
libmypaint_gegl_la_LIBS = $(GEGL_LIBS) ../libmypaint.la
```
in `gegl/Makefile.am`, file `gegl/.libs/libmypaint-gegl.so` does not end up being linked against GEGL or libmypaint, not without gold, not without as-needed, never. Why? See:
```
# ./configure --enable-introspection --enable-gegl && make clean all
[..]
# lddtree gegl/.libs/libmypaint-gegl.so
libmypaint-gegl.so => gegl/.libs/libmypaint-gegl.so (interpreter => none)
    libm.so.6 => /lib64/libm.so.6
        ld-linux-x86-64.so.2 => /lib64/ld-linux-x86-64.so.2
    libpthread.so.0 => /lib64/libpthread.so.0
    libc.so.6 => /lib64/libc.so.6
```
That case doesn't error out though.

**Comment by achadwick** [comment]
Closed by 94f059dbeaa94991466964d0357e6f6e54bb5ece.
